Application/ControlMsrE2/ControlMsrE2.c

```diff
@@ -0,0 +1,252 @@
+          IfrHeader = PADD (PkgHeader, sizeof (EFI_HII_PACKAGE_HEADER));
+          //
+          // Form Definition must start with FORM_SET_OP
+          //
+          if (IfrHeader->OpCode == EFI_IFR_FORM_SET_OP) {
+            DEBUG ((
+              DEBUG_INFO,
+              "Form: %g\n",
+              &((EFI_IFR_FORM_SET *) IfrHeader)->Guid
+              ));
+
+            if (IfrHeader->Length >= 16 + sizeof (EFI_IFR_FORM_SET)) {
+              DEBUG ((
+                DEBUG_INFO,
+                "Class Guid: %g\n",
+                PADD (IfrHeader, sizeof (EFI_IFR_FORM_SET))
+                ));
+
+              //
+              // Checkup for Setup Form
+              //
+              if (CompareGuid (&gEfiHiiPlatformSetupFormsetGuid, PADD (IfrHeader, sizeof (EFI_IFR_FORM_SET)))) {
+                Contexts[ContextsCount].SearchText = SearchString;
+                Contexts[ContextsCount].EfiHandle = HiiHandles[ListHeaderIndex];
+                Contexts[ContextsCount].ListHeader = ListHeaders[ListHeaderIndex];
+                Contexts[ContextsCount].PkgHeader = PkgHeader;
+                Contexts[ContextsCount].FirstIfrHeader = PADD (IfrHeader, IfrHeader->Length);
+                Contexts[ContextsCount].IfrVarStore = NULL;
+                Contexts[ContextsCount].IfrOneOf = NULL;
+                Contexts[ContextsCount].StopAt = DONT_STOP_AT;
+                Contexts[ContextsCount].Count = OptionsCount;
+
+                IterateOpCode (
+                  Contexts[ContextsCount].FirstIfrHeader,
+                  EFI_IFR_ONE_OF_OP,
+                  NULL,
+                  &Contexts[ContextsCount],
+                  HandleIfrOption
+                  );
+
+                if (Contexts[ContextsCount].Count != OptionsCount) {
+                  OptionsCount = Contexts[ContextsCount].Count;
+                  ++ContextsCount;
+                }
+              }
+            }
+          }
+        }
+        PkgHeader = PADD (PkgHeader, PkgHeader->Length);
+      }  ///< For each package in list
+
+      DEBUG ((DEBUG_INFO, "\n"));
+    }  ///< ListHeader End
+  }  ///< For Each Handle
+
+  DEBUG ((DEBUG_INFO, "Context Count: %x Options Count %x\n", ContextsCount, OptionsCount));
+
+  if (IS_INTERACTIVE ()
+   || IS_LOCK ()
+   || IS_UNLOCK ()) {
+    if (OptionsCount > 9 || ContextsCount == CONTEXTS_MAX) {
+      Print (L"Too many corresponding BIOS Options found. Try a different search string using interactive mode.\n");
+    } else if (OptionsCount == 0) {
+      Print (L"No corresponding BIOS Options found. Try a different search string using interactive mode.\n");
+    } else {
+      Key = '1';
+
+      if (OptionsCount > 1) {
+        do {
+          Print (L"\nEnter choice (1..%x) ? ", OptionsCount);
+          Key = ReadAnyKey ();
+        } while ((Key < '1') && (Key > '0' + OptionsCount) && (Key != 0x1B));
+
+        Print (L"\n");
+      }
+
+      if (Key != 0x1B) {
```

Comment:
I know there is no public definition for this, but maybe replicate EDK II's CHAR_ESC.